# Customer Retention Analysis

## Libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import plotly.express as px

from nb_pyscript.data import data_overview

Basic settings of dataframe and plots

In [2]:
# Pandas dataframes show all rows and columns without skiping
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

# Matplotlib setting of figsize and dpi
mpl.rcParams['figure.figsize'] = [16, 5]
mpl.rcParams['figure.dpi'] = 200

%matplotlib inline

## Dataset

In [4]:
data = pd.read_excel('./data/telco-customer-churn/telco_customer_churn.xlsx')

## Data Exploration

In [5]:
df = data.copy()

In [45]:
data_overview(df, "Overview of telco customer data:")

Overview of telco customer data:
Entries in the dataset: 7043
Features in the dataset: 33


Faatures of the dataset:
 ['CustomerID', 'Count', 'Country', 'State', 'City', 'Zip Code', 'Lat Long', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen', 'Partner', 'Dependents', 'Tenure Months', 'Phone Service', 'Multiple Lines', 'Internet Service', 'Online Security', 'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV', 'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method', 'Monthly Charges', 'Total Charges', 'Churn Label', 'Churn Value', 'Churn Score', 'CLTV', 'Churn Reason']


Missing Values:
CustomerID                    : 0
Count                         : 0
Country                       : 0
State                         : 0
City                          : 0
Zip Code                      : 0
Lat Long                      : 0
Latitude                      : 0
Longitude                     : 0
Gender                        : 0
Senior Citizen                : 0

**About the dataset:**

- Dataset contains 7043 entries and 33 features.

- Features of the dataset are-
    - `Customer ID`: ID of customers each are unique
    - `Count`: Frequency of the customer in the table
    - `Country`: Country of the customer
    - `State`: State where customer lives
    - `City`: City where customer resides(1129 unique city names)
    - `Zip Code`: Zip code of the place
    - `Lat Long`: Latitude and Longitude of the place
    - `Latitude`: Latitude of the place
    - `Longitude`: Longitude of the place
    - `Gender`: Gender of the customer (Male or Female)
    - `SeniorCitizen`: Indicates whether the client is an older person (0, 1).
    - `Partner`: Indicates whether or not the client is partnered (Yes, No).
    - `Dependents`: Indicates whether the client is supported by others (Yes, No).
    - `Tenure Months`: The length of time a customer has been a customer of the business (multiple different number values)).
    - `Phone Service`: If the client has a phone service, it is indicated by the words “Yes” or “No”.
    - `Multiple Line`: Whether the customer has more than one line (no phone service, no service, yes service).
    - `Internet Service`: Whether the client has a subscription to the company’s Internet service (DSL, Fiber optic, or No).
    - `Online Security`: Indicates if the client has access to online security (Internet service available, No, Yes).
    - `Online Backup`: Indicates whether or not the client has an online backup (Internet service unavailable, No, Yes).
    - `Device Protection`: Indicates whether the client has device protection (Internet service not available, Not Available, Yes).
    - `Tech Support`: Whether the customer has access to tech help (no internet service, no, yes).
    - `Streaming TV`: Whether the customer has access (no internet service, no, yes).
    - `Streaming Movies`: Streaming movies: Indicates whether or not the client offers or has access streaming movies(no internet service, no, yes).
    - `Contract`: The type of existing contract for the customer (Month-to-Month, One-Year, Two-Year)
    - `Paperless Billing`: Whether the client uses paperless billing (Yes, No)
    - `PaymentMethod`: The chosen payment method by the consumer (credit card, bank transfer, electronic check, paper check)
    - `Monthly Charges`: The monthly charge made to the consumer (various numeric quantities)
    - `Total Charges`: (many different numeric values): The total amount charged to the consumer
    - `Churn Label`: Whether the customer decided to leave the service or not ('yes' or 'no').
    - `Churn Value`: 0 or 1 depending on the churn value as 'no' or 'yes'.
    - `CLTV`: Customer Lifetime Value if they stayed in the company as a customer.
    - `Churn Reason`: Churn Reason: Reason for the customer to choose to leave.sumer. No).